src/ompl/base/samplers/InformedStateSampler.h

```diff
@@ -0,0 +1,109 @@
+/*********************************************************************
+* Software License Agreement (BSD License)
+*
+*  Copyright (c) 2014, University of Toronto
+*  All rights reserved.
+*
+*  Redistribution and use in source and binary forms, with or without
+*  modification, are permitted provided that the following conditions
+*  are met:
+*
+*   * Redistributions of source code must retain the above copyright
+*     notice, this list of conditions and the following disclaimer.
+*   * Redistributions in binary form must reproduce the above
+*     copyright notice, this list of conditions and the following
+*     disclaimer in the documentation and/or other materials provided
+*     with the distribution.
+*   * Neither the name of the University of Toronto nor the names of its
+*     contributors may be used to endorse or promote products derived
+*     from this software without specific prior written permission.
+*
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+
+/* Authors: Jonathan Gammell */
+
+#ifndef OMPL_BASE_SAMPLERS_INFORMED_SAMPLER_
+#define OMPL_BASE_SAMPLERS_INFORMED_SAMPLER_
+
+//We inherit from StateSampler
+#include "ompl/base/StateSampler.h"
+//Deriving functions must be able to sample within a given cost
+#include "ompl/base/Cost.h"
+//We use a pointer to the problem definition to access problem and solution data.
+#include "ompl/base/ProblemDefinition.h"
+
+namespace ompl
+{
+    namespace base
+    {
+        OMPL_CLASS_FORWARD(InformedStateSampler);
+
+        /** \brief An abstract class for state samplers that use information
+        about the current solution to limit future search to a planning
+        subproblem that contains all possibly better solutions. */
+        class InformedStateSampler : public StateSampler
+        {
+        public:
+            /** \brief Construct a sampler that only generates states with a heuristic solution estimate that is less than the cost of the current solution. */
+            InformedStateSampler(const StateSpace* space, const ProblemDefinitionPtr probDefn, const Cost* bestCost);
```

Comment:
First, a bit of background: ValidStateSampler is conceptually different than a StateSampler in the following way. A StateSampler implements a means of sampling a space, with some probability distribution. As such, it can never ever fail. When called, it will certainly produce a sample within the given pdf. A ValidStateSampler however is a wrapper that only samples a subset of the state space, a subset which may be empty or sufficiently small that we cannot find it with every call. Due to this difference we opted for different classes.

Now, the InformedStateSampler -- does it fall under the description above? Is it always the case generating a sample makes sense / there exists no failure mode? From the implementation it seems that may be the case, but the cost guarantees may not be there.